File: most_water/most_water.go
// https://leetcode.com/problems/container-with-most-water/description/
package most_water

func maxArea(heights []int) int {
	i := 0
	j := len(heights) - 1
	areaMax := 0

	for i < j {
		area := (j - i) * min(heights[i], heights[j])
		if area > areaMax {
			areaMax = area
		}

		if heights[i] <= heights[j] {
			i++
		} else {
			j--
		}
	}
	return areaMax
}

File: most_water/most_water.py
"""https://leetcode.com/problems/container-with-most-water/description/"""


def maxArea(heights: list[int]) -> int:
    left_idx = 0
    right_idx = len(heights) - 1
    max_area = 0
    while left_idx < right_idx:
        left_height = heights[left_idx]
        right_height = heights[right_idx]

        area = (right_idx - left_idx) * min(left_height, right_height)
        if area > max_area:
            max_area = area

        if left_height <= right_height:
            left_idx += 1
        else:
            right_idx -= 1

    return max_area
